### PR TITLE
Virtual counters

### PR DESCRIPTION
**Pull request by hexoscott**
Early eyes for now on the general approach

### CODE REVIEW

File: core/vm/contract.go [hunk 1]
```diff
@@ -58,6 +58,9 @@ type Contract struct {
 
 	Gas   uint64
 	value *uint256.Int
+
+	// zkevm
+	IsCreate bool
```
Comment (revitteth) [inline core/vm/contract.go:63]: contract_zkevm?

File: core/vm/interpreter.go [hunk 2]
```diff
@@ -28,22 +28,6 @@ import (
 	"github.com/ledgerwatch/erigon/core/vm/stack"
 )
 
-// Config are the configuration options for the Interpreter
-type Config struct {
```
Comment (revitteth) [inline core/vm/interpreter.go]: Possibly ConfigZk in the new file so we avoid a conflict with the delete here?

File: core/vm/interpreter.go [hunk 3]
```diff
@@ -28,22 +28,6 @@ import (
 	"github.com/ledgerwatch/erigon/core/vm/stack"
 )
 
-// Config are the configuration options for the Interpreter
```
Comment (mandrigin) [inline core/vm/interpreter.go]: why removing that?

File: core/vm/interpreter_zkevm.go [hunk 4]
```diff
@@ -1,6 +1,28 @@
 package vm
 
-import "github.com/ledgerwatch/log/v3"
+import (
+	"github.com/ledgerwatch/erigon/zk/sequencer"
+	"github.com/ledgerwatch/log/v3"
+)
+
+// Config are the configuration options for the Interpreter
+type Config struct {
```
Comment (mandrigin) [inline core/vm/interpreter_zkevm.go]: maybe we can have a wrapper struct `ZkConfig { config: Config collector: *CounterCollector }` ?

File: core/vm/interpreter_zkevm.go [hunk 5]
```diff
@@ -14,8 +36,8 @@ func NewZKEVMInterpreter(evm VMInterpreter, cfg Config) *EVMInterpreter {
 	case evm.ChainRules().IsBerlin:
 		jt = &zkevmForkID4InstructionSet
 	}
+	jt = copyJumpTable(jt)
```
Comment (mandrigin) [inline core/vm/interpreter_zkevm.go:39]: why this change?

File: core/vm/zk_batch_counters.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+		blockCount:   0,
+	}
+}
+
+// AddNewTransactionCounters makes the collector aware that a new transaction is attempting to be added to the collector
+// here we check the batchL2Data length and ensure that it doesn't cause an overflow.  This will be re-calculated
+// every time a new transaction is added as it needs to take into account all the transactions in a batch.
+// The function will return false in the case of an error or an overflow of counters
+func (bcc *BatchCounterCollector) AddNewTransactionCounters(txCounters *TransactionCounter) (bool, error) {
+	bcc.transactions = append(bcc.transactions, txCounters)
```
Comment (mandrigin) [inline core/vm/zk_batch_counters.go:32]: hmm, what happens if it overflows? bcc.transactions stays the same?

Comment (mandrigin) [reply core/vm/zk_batch_counters.go:32]: here and everywhere -- won't there be an extra operation?

File: eth/backend.go
Comment (V-Staykov) [inline eth/backend.go:1]: in etrog branch we've separated this into a new file and would need to rebase the changes. But since you are making changes to it here, maybe separate it here as well? We've basically just copied `ExecuteBlockEphemerally` into the new file and renamed it `ExecuteBlockEphemerallyZk`

File: core/vm/interpreter_zkevm.go (hunk 5 above)
Comment (hexoscott) [reply core/vm/interpreter_zkevm.go:39]: I need to clone the original jump table in every case otherwise we end up wrapping and wrapping each time this function is called and get a deeper stack with multiple iterations of the counters being handled

File: core/vm/zk_batch_counters.go (hunk 6 above)
Comment (hexoscott) [reply core/vm/zk_batch_counters.go:32]: So the thought is that we only care for counters until they overflow, once it has or we reach the time limit we don't need to hold on to them.  So although the transactions slice will contain the transaction that caused the overflow it doesn't matter because we won't be including that last transaction in the block

File: core/vm/interpreter_zkevm.go (hunk 4 above)
Comment (hexoscott) [reply core/vm/interpreter_zkevm.go]: Good idea, saves some confusion, made the change, will push shortly

File: core/vm/contract.go (hunk 1 above)
Comment (hexoscott) [reply core/vm/contract.go:63]: I can take a look but I think this would create wider spread changes rippling through the codebase.  I know we want to avoid clashes with upstream but there is a one liner to set this in the contract create flow otherwise it's left to default as false

File: core/vm/zk_batch_counters.go (hunk 6 above)
Comment (hexoscott) [reply core/vm/zk_batch_counters.go:32]: If you check in the stage_execution code you'll see we add a tx where this is called, then check for an overflow immediately, then execute, then check for an overflow post execution.  Only if it hasn't overflowed at these stages do we add the TX into the slice ready to be included in the block. If there's an overflow we quit early and process the block without this TX

File: core/vm/contract.go (hunk 1 above)
Comment (mandrigin) [reply core/vm/contract.go:63]: fair enough

File: core/vm/evm.go [hunk 7]
```diff
@@ -103,7 +103,11 @@ func NewEVM(blockCtx evmtypes.BlockContext, txCtx evmtypes.TxContext, state evmt
 	}
 
 	// [zkevm] change
-	evm.interpreter = NewZKEVMInterpreter(evm, vmConfig)
+	zkConfig := ZkConfig{
```
Comment (mandrigin) [inline core/vm/evm.go]: to keep it more rebase-friendly, let's have a method like `NewZkConfig` and use it directly in the `evm.interpreter` assignment like that `evm.interpreter = NewZKEVMInterpreter(evm, NewZKConfig(vmConfig))` here and in another place.

File: core/vm/interpreter.go [hunk 8]
```diff
@@ -28,6 +28,12 @@ import (
 	"github.com/ledgerwatch/erigon/core/vm/stack"
 )
 
+var pool = sync.Pool{
```
Comment (mandrigin) [inline core/vm/interpreter.go]: let's not move this code, again, for the sake of rebases

File: core/vm/interpreter_zkevm.go (hunk 5 above)
Comment (mandrigin) [reply core/vm/interpreter_zkevm.go:39]: please leave a comment then :)

File: core/vm/zk_batch_counters.go (hunk 6 above)
Comment (mandrigin) [reply core/vm/zk_batch_counters.go:32]: okay got it

**Review by mandrigin — CHANGES_REQUESTED** [review]
very minor things to fix

**Comment by hexoscott** [comment]
Changes made @mandrigin

**Review by mandrigin — APPROVED** [review]
